examples/mnist/dataset.go

```diff
@@ -0,0 +1,304 @@
+	width               = 28
+	height              = 28
+	numClasses          = 10
+	trainExamples       = 60000
+	testExamples        = 10000
+
+	imageMagic = 0x00000803
+	labelMagic = 0x00000801
+)
+
+type fileType int
+
+const (
+	imageFileType fileType = iota
+	labelFileType
+)
+
+var mnistFiles = map[string][2]string{
+	"train": {trainImagesFilename, trainLabelsFilename},
+	"test":  {testImagesFilename, testLabelsFilename},
+}
+
+var mnistSamples = map[string]int{
+	"train": trainExamples,
+	"test":  testExamples,
+}
+
+// Image represents a MNIST image. It is a array a bytes representing the color.
+// 0 is black (the background) and 255 is white (the digit color).
+type Image [width * height]byte
+
+// Label is the digit label from 0 to 9.
+type Label = int8
+
+type imageFileHeader struct {
+	Magic     int32
+	NumImages int32
+	Height    int32
+	Width     int32
+}
+
+type labelFileHeader struct {
+	Magic     int32
+	NumLabels int32
+}
+
+// ColorModel implements the image.Image interface.
+func (img Image) ColorModel() color.Model {
+	return color.GrayModel
+}
+
+// Bounds implements the image.Image interface.
+func (img Image) Bounds() image.Rectangle {
+	return image.Rectangle{
+		Min: image.Point{0, 0},
+		Max: image.Point{width, height},
+	}
+}
+
+// At implements the image.Image interface.
+func (img Image) At(x, y int) color.Color {
+	return color.Gray{Y: img[y*width+x]}
+}
+
+// Set modifies the pixel at (x,y).
+func (img *Image) Set(x, y int, v byte) {
+	img[y*width+x] = v
+}
+
+// Download MNIST Dataset to baseDir, unzips it
+func Download(baseDir string) error {
+	baseDir = data.ReplaceTildeInDir(baseDir)
+	files := []string{trainImagesFilename, trainLabelsFilename, testImagesFilename, testLabelsFilename}
+	for _, file := range files {
+		downloadURLFile, _ := url.JoinPath(downloadURL, file)
+		filePath := path.Join(baseDir, file)
+		if err := data.DownloadIfMissing(downloadURLFile, filePath, ""); err != nil {
+			return fmt.Errorf("data.DownloadAndUnzipIfMissing: %w", err)
+		}
+	}
+
+	return nil
+}
+
+var (
+	AssertImageIsImageImage *Image
+	_                       image.Image = AssertImageIsImageImage
+)
+
+// NewDataset creates a train.Dataset that yields images from Dogs vs Cats Dataset.
+//
+// It takes the following arguments:
+//
+//   - name:
+//   - baseDir:
+//   - mode: choose between 'train' and 'test'
+func NewDataset(backend backends.Backend, name, baseDir, mode string, dtype dtypes.DType) (ds *data.InMemoryDataset, err error) {
+	imagesFile := mnistFiles[mode][imageFileType]
+	labelsFile := mnistFiles[mode][labelFileType]
+
+	images, err := loadImageFile(path.Join(baseDir, imagesFile))
+	if err != nil {
+		return nil, err
+	}
+	labels, err := loadLabelFile(path.Join(baseDir, labelsFile))
+	if err != nil {
+		return nil, err
+	}
+
+	return data.InMemoryFromData(
+		backend,
+		name,
+		[]any{timage.ToTensor(dtype).Batch(images)},
+		[]any{tensors.FromAnyValue(shapes.CastAsDType(oneHot(labels, numClasses), dtype))},
```

Comment:
I don't think you need to `oneHot` the label here.

During the calculation of the loss it will need to "one-hot" the label. But you can keep it encoded as an int (there is even a `OneHot` function defined in `graph`).

Notice that multi-class losses like `losses.CategoricalCrossEntropy{Logits}` take the labels in "dense" format, that is, the label is a number from 0 to numClasses-1.

This changes when instead of having one correct class you have as labels a probability distribution over classes. For that there is the `losses.SparseCategoricalCrossEntropyLogits`. But that is not the case here. In MNIST each example is labeled as one, and only one, digit.